Enable Object Detail Visualization in Questions and Dashboards

iethree opened the pull request:
## Description

Allows our object detail visualization to be used, not just in a modal, but as a normal visualization in questions and dashboards. For questions with more than one result, we show a little paginator to allow scrolling between records

Works for Questions

![Screen Shot 2023-01-18 at 4 32 14 PM](https://user-images.githubusercontent.com/30528226/213318146-a8d5d4c9-cd39-485c-8089-ab7fdbaa6148.png)

Works for Dashboards

![Screen Shot 2023-01-18 at 4 27 37 PM](https://user-images.githubusercontent.com/30528226/213318037-54a62eac-2e09-4949-8728-44cd13be2bf8.png)

Paginates result sets with multiple rows

![pagination](https://user-images.githubusercontent.com/30528226/213318022-a5decbd0-951e-4bbd-bead-0239ddff32ba.gif)

iethree committed:
Allow questions to use Object Detail Visualization

iethree committed:
add unit tests for new functionality

alxnddr reviewed (commented):
Looks good, I got a couple of findings and questions:

1) Should we hide the column settings? They don't change the order of values, adding new columns does not add rows:
 
<img width="1153" alt="Screen Shot 2023-01-19 at 2 29 10 PM" src="https://user-images.githubusercontent.com/14301985/213517298-9bbe809b-fed9-4a42-a185-f58226eae5cc.png">

2) Click behavior does not work. Also, how exactly it should work, what should be clickable?
<img width="1728" alt="Screen Shot 2023-01-19 at 2 32 05 PM" src="https://user-images.githubusercontent.com/14301985/213517966-ca2e8ec3-ba66-48dc-a3e9-4c9921616a6d.png">

3) The button in the right corner does exactly the same as clicking on the card title. Is there anything specific it should perform? Also, both clicks lead to the QB showing the very first row even though on a dashboard I multiple times clicked "next item" button on the object detail card. This can be misleading 

<img width="1186" alt="Screen Shot 2023-01-19 at 2 33 45 PM" src="https://user-images.githubusercontent.com/14301985/213518279-318f1805-21be-4303-a804-f32f20f50e16.png">

iethree commented:
> Looks good, I got a couple of findings and questions:
> 
> 1. Should we hide the column settings? They don't change the order of values, adding new columns does not add rows:
> 
> <img alt="Screen Shot 2023-01-19 at 2 29 10 PM" width="1153" src="https://user-images.githubusercontent.com/14301985/213517298-9bbe809b-fed9-4a42-a185-f58226eae5cc.png">
> 
> 2. Click behavior does not work. Also, how exactly it should work, what should be clickable?
> 
> <img alt="Screen Shot 2023-01-19 at 2 32 05 PM" width="1728" src="https://user-images.githubusercontent.com/14301985/213517966-ca2e8ec3-ba66-48dc-a3e9-4c9921616a6d.png">
> 
> 3. The button in the right corner does exactly the same as clicking on the card title. Is there anything specific it should perform? Also, both clicks lead to the QB showing the very first row even though on a dashboard I multiple times clicked "next item" button on the object detail card. This can be misleading
> 
> <img alt="Screen Shot 2023-01-19 at 2 33 45 PM" width="1186" src="https://user-images.githubusercontent.com/14301985/213518279-318f1805-21be-4303-a804-f32f20f50e16.png">

thanks sasha, those settings changes were working in `data-apps-main`, let me see what the cause seems to be.

iethree committed:
use column settings in object detail table

iethree committed:
remove questionLink icon

iethree committed:
disable click behavior for object detail

iethree commented:
1. Column ordering and add/remove now works properly. I forgot to import updates to the objectDetailTable file. I also found and fixed an interesting little bug there.

![reorderCols](https://user-images.githubusercontent.com/30528226/213809381-c6be0d48-5c0d-4734-acbf-65f11b66e6a6.gif)

2. I think the right thing to do is just disable click behavior for this viz for now until we can thoroughly think through what expected click behavior might be.

![Screen Shot 2023-01-20 at 2 30 57 PM](https://user-images.githubusercontent.com/30528226/213809532-4a577278-9a43-4075-9e06-d0ca7c63f419.png)

3. I removed the "insight" icon - that was put there for use in data apps where cards didn't have titles - which no longer matters.

iethree commented on frontend/src/metabase/visualizations/components/settings/ChartSettingOrderedColumns.jsx:
```diff
@@ -28,7 +28,9 @@ export default class ChartSettingOrderedColumns extends Component {
 
   handleSortEnd = ({ oldIndex, newIndex }) => {
-    const fields = [...this.props.value];
+    const fields = [...this.props.value].sort((a, b) =>
+      a.enabled && !b.enabled ? -1 : 1,
+    );
```
@npfitz and @alxnddr , I would appreciate your eyes on this line. This was necessary because column reordering didn't work properly for object detail viz because the `value` prop here contained all columns, even disabled columns. This behaves differently for table viz, because we re-run the query every time we show or hide a column (which is probably not ideal), and that removes them from this value array. 

![Screen Shot 2023-01-20 at 2 39 53 PM](https://user-images.githubusercontent.com/30528226/213810035-d5917264-b47e-4ca1-a1ea-51655e163c3a.png)


The behavior i was seeing was that the enabled columns list would have, for instances, indexes `0,6,7` when columns 1-5 were hidden, but the ordering component would re-index those as `0,1,2`. So if you wanted to move the first item to the end of the list, it would move from position 0 -> 2 ... which would still order at the beginning because the objectDetail component still things there are 8 columns.

By making sure that `enabled` columns are always first, we ensure that we're properly sorting the order only amongst enabled columns and we get the intended behavior. I've also tested it on the table component, and it seems to work properly there as well. It also seems to work properly with pivot tables.

Nonetheless, I'd appreciate you throwing any red flags you see here, or maybe suggesting an alternative approach if you can think of one.

npfitz replied:
Uh oh, one red flag I can see is that I'm removing that settings component in my PR (though I don't necessarily need to). Maybe we can sync up and see if there are alternatives

iethree replied:
This has now been updated to use the new `ChartSettingsColumnEditor` and `ChartSettingOrderedSimple` components

npfitz replied:
Woo!

iethree commented:
🧊  this is on ice for the moment so it can take advantage of some new column ordering components in https://github.com/metabase/metabase/issues/25453

iethree committed:
use new column settings widgets for object detail settings

iethree committed:
update tests

iethree committed:
update nulls e2e test

alxnddr commented:
Seems like we need to adjust column settings here to exclude columns from related tables:
<img width="1081" alt="Screen Shot 2023-02-14 at 1 40 51 PM" src="https://user-images.githubusercontent.com/14301985/218800924-f994cf50-5699-4d73-ba09-fe8bcac05a13.png">

alxnddr reviewed (commented):
#### It would be nice to improve the empty state:
<img width="939" alt="Screen Shot 2023-02-14 at 1 42 04 PM" src="https://user-images.githubusercontent.com/14301985/218801300-7cc34700-3441-4183-8338-7932f319f7d2.png">

#### We should apply formatting to the header value:

In this screenshot, 113 is a product ID that is remapped to EAN (0212722801067), so the header should also show 0212722801067
<img width="608" alt="Screen Shot 2023-02-14 at 1 48 13 PM" src="https://user-images.githubusercontent.com/14301985/218802753-64e1d798-8738-4554-b9ea-a18cb628dbf6.png">

It is more illustrative with dates:
<img width="598" alt="Screen Shot 2023-02-14 at 1 45 29 PM" src="https://user-images.githubusercontent.com/14301985/218803132-4afec154-0a38-41c8-89cb-0a25909316fa.png">

alxnddr reviewed (commented):
Besides a few minor comments I left, looks great!

iethree commented:
> Seems like we need to adjust column settings here to exclude columns from related tables:

This actually used to work with the old column picker, and sort of does still if you add the columns in table view and THEN switch to detail view. I'll take a closer look.

![Screen Shot 2023-02-15 at 8 48 40 AM](https://user-images.githubusercontent.com/30528226/219078806-78cb853d-ad3d-4275-8741-4451fde326c6.png)

iethree committed:
better empty state for missing columns

iethree committed:
obey column settings in object detail header

iethree committed:
update tests

iethree committed:
use table.column setting

iethree commented:
> Seems like we need to adjust column settings here to exclude columns from related tables

@alxnddr I think excluding linked tables isn't the right solution here because this also causes problems if you have a query with less-than-all of the primary table's columns and try yo add one back.

Instead, I switched this viz to share the `table.columns` setting. This made everything work properly and re-runs the query appropriately when additional columns are added. I also think, from a ux perspective, that people will expect to see the same columns when switching back/forth between detail and table view.

![Screen Shot 2023-02-16 at 2 22 41 PM](https://user-images.githubusercontent.com/30528226/219490667-b396e10b-0c94-4164-a7b6-0832f8b0c538.png)

iethree committed:
fix click map mesasge
